prov/gni/test/cntr.c

```diff
@@ -797,3 +797,89 @@ Test(cntr_mt, write_wait)
 
 	dbg_printf("done\n");
 }
+
+void *do_add_cntr(void *arg)
+{
+	int i = 0, ret, iters = ((int *)arg)[0];
+
+	for (; i < iters; i++) {
+		ret = fi_cntr_add(ep_write_cntrs[0], 1);
+		cr_assert(ret == FI_SUCCESS);
+	}
+
+	return NULL;
+}
+
+void *do_add_err_cntr(void *arg)
+{
+	int i = 0, ret, iters = ((int *)arg)[0];
+
+	for (; i < iters; i++) {
+		ret = fi_cntr_adderr(ep_write_cntrs[0], 1);
+		cr_assert(ret == FI_SUCCESS);
+	}
+
+	return NULL;
+}
+
+Test(cntr_mt, set_add_read_cntr)
+{
+	int iters = 128, nthreads = 4, i, ret;
```

Comment:
If you want to test the atomicity of the function, you'd need to do broader locking, which would be pointless because access to the function has been serialized. 

Add operations are not going to be fully atomic, because they involve multiple instructions that do not lock memory for the entirety of the function call. Any thread could be put to sleep between the invocation of the call and the call to atomic_add. Any subsequent call to read on the counter value could be proceeded by an atomic add from a different context, thus making your result invalid w.r.t. to your expected value. 

You can't verify that the function is atomic without checking the contents of the variable prior and after each call to the atomic itself. This is probably best tested in a different suite that exercises the atomics themselves, rather than the counters.